src/cascadia/PublicTerminalCore/Source.cpp

```diff
@@ -0,0 +1,291 @@
+}
+
+void HwndTerminal::_UpdateFont(int newDpi)
+{
+    auto lock = _terminal->LockForWriting();
+
+    // TODO: MSFT:20895307 If the font doesn't exist, this doesn't
+    //      actually fail. We need a way to gracefully fallback.
+    _renderer->TriggerFontChange(newDpi, _desiredFont, _actualFont);
+}
+
+HRESULT HwndTerminal::Refresh(double width, double height, _Out_ int* charColumns, _Out_ int* charRows)
+{
+    auto lock = _terminal->LockForWriting();
+    const SIZE windowSize{ static_cast<short>(width), static_cast<short>(height) };
+    RETURN_IF_FAILED(_renderEngine->SetWindowSize(windowSize));
+
+    // Invalidate everything
+    _renderer->TriggerRedrawAll();
+
+    // Convert our new dimensions to characters
+    const auto viewInPixels = Viewport::FromDimensions({ 0, 0 },
+                                                       { static_cast<short>(windowSize.cx), static_cast<short>(windowSize.cy) });
+    const auto vp = _renderEngine->GetViewportInCharacters(viewInPixels);
+
+    // If this function succeeds with S_FALSE, then the terminal didn't
+    //      actually change size. No need to notify the connection of this
+    //      no-op.
+    // TODO: MSFT:20642295 Resizing the buffer will corrupt it
+    // I believe we'll need support for CSI 2J, and additionally I think
+    //      we're resetting the viewport to the top
+    RETURN_IF_FAILED(_terminal->UserResize({ vp.Width(), vp.Height() }));
+    *charColumns = vp.Width();
+    *charRows = vp.Height();
+
+    return S_OK;
+}
+
+void HwndTerminal::SendOutput(std::wstring_view data)
+{
+    _terminal->Write(data);
+}
+
+HRESULT _stdcall CreateTerminal(HWND parentHwnd, _Out_ void** hwnd, _Out_ void** terminal)
+{
+    auto _terminal = new HwndTerminal(parentHwnd);
+    RETURN_IF_FAILED(_terminal->Initialize());
+
+    *terminal = _terminal;
+    *hwnd = _terminal->_hwnd;
+
+    return S_OK;
+}
+
+void _stdcall RegisterScrollCallback(void* terminal, void __stdcall callback(int, int, int))
+{
+    auto publicTerminal = (HwndTerminal*)terminal;
+    publicTerminal->RegisterScrollCallback(callback);
+}
+
+void _stdcall SendTerminalOutput(void* terminal, LPCWSTR data)
+{
+    auto publicTerminal = (HwndTerminal*)terminal;
+    publicTerminal->SendOutput(data);
+}
+
+HRESULT _stdcall TriggerResize(void* terminal, double width, double height, _Out_ int* charColumns, _Out_ int* charRows)
+{
+    auto publicTerminal = (HwndTerminal*)terminal;
+    return publicTerminal->Refresh(width, height, charColumns, charRows);
+}
+
+void _stdcall DpiChanged(void* terminal, int newDpi)
+{
+    auto publicTerminal = (HwndTerminal*)terminal;
+    publicTerminal->_UpdateFont(newDpi);
+}
+
+void _stdcall UserScroll(void* terminal, int viewTop)
+{
+    auto publicTerminal = (HwndTerminal*)terminal;
+    publicTerminal->_terminal->UserScrollViewport(viewTop);
+}
+
+void _stdcall StartSelection(void* terminal, COORD cursorPosition, bool altPressed)
+{
+    COORD terminalPosition = {

```

Comment:
Sorry, what? Why is there casting at all here from the same type to the same type?